projects/stretch_continual/stretch_continual/envs/stretch_demo_base_env.py

```diff
@@ -0,0 +1,312 @@
+import math
+import os
+import tempfile
+import time
+
+import gym
+import h5py
+import numpy as np
+from data_tools.image import img_from_bytes
+
+from home_robot.motion.stretch import HelloStretchIdx
+
+CAMERA_FAR_PLANE = 1  # m
+CAMERA_SCALE_CONST = 10000
+STRETCH_URDF_DIR = os.environ["STRETCH_URDF_DIR"]
+
+
+class CacheContainer(object):
+    """
+    Spoof file-like object for when we want to cache the h5 information instead of re-loading it
+    """
+
+    def __init__(self, cached_data):
+        self.cached_data = cached_data
+
+    def __enter__(self):
+        return self.cached_data
+
+    def __exit__(self, *args):
+        pass
+
+
+class StretchDemoBaseEnv(gym.Env):
+    # The URDFs provided with stretch_control do not support base_x_joint motion, so ignore that during IK for now
+    MANIP_MODE_CONTROLLED_JOINTS = [
+        "ignore",  # Originally base_x_joint
+        "joint_lift",
+        "joint_arm_l3",
+        "joint_arm_l2",
+        "joint_arm_l1",
+        "joint_arm_l0",
+        "joint_wrist_yaw",
+        "joint_wrist_pitch",
+        "joint_wrist_roll",
+    ]
+
+    # This EE link does not require any offsets or conversions to use --
+    # it is located directly in the middle of the gripper
+    EE_LINK_NAME = "link_gripper_fingertip_center"
+
+    def __init__(self, initialize_ros, include_context):
+        super().__init__()
+        self._urdf_path = STRETCH_URDF_DIR  # os.path.join(STRETCH_URDF_DIR, "planner_calibrated_manipulation_mode.urdf")  # TODO: pass in urdf path
+        self._initialize_ros = initialize_ros  # TODO: remove?
+        self._include_context = include_context
+        self._trajectory_cache = {"linked_files": {}}
+
+    def _recursive_listdir(self, directory):
+        # From: https://stackoverflow.com/questions/19309667/recursive-os-listdir
+        return [os.path.join(dp, f) for dp, dn, fn in os.walk(directory) for f in fn]
+
+    def _recursively_load_h5py(self, trajectory_data):
+        extracted_data = {}
+        for traj_key, traj_value in trajectory_data.items():
+            if isinstance(traj_value, h5py.Group):
+                extracted_data[traj_key] = self._recursively_load_h5py(traj_value)
+            else:
+                extracted_data[traj_key] = np.array(traj_value)
+
+        return extracted_data
+
+    def load_all_h5_from_dir(
+        self, directory, only_key_frames, temp_aggregation_file, cache
+    ):
+        """
+        We may have multiple h5 files, each of which may have multiple trajectories. This method loads all the h5 files
+        into an aggregated h5 via ExternalLink. If we're caching, we'll load it into a local dictionary, and spoof
+        that as a file-like object, so we can use these loading methods interchangeably.
+        """
+        h5_id = 0
+        output_path = temp_aggregation_file.name
+        max_tries = 5
+
+        if not cache or len(self._trajectory_cache["linked_files"]) == 0:
+            for try_id in range(max_tries):
```

Comment:
what is this expected to do? why would this need a try?

Reply:
I was experiencing issues with multiprocessing, with multiple files accessing the same h5. Let me see if it's still an issue....